onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py

```diff
@@ -903,7 +903,15 @@ def _infer_Scan(self, node):
         scan_input_axes = get_attribute(node, 'scan_input_axes', [0]*num_scan_inputs)
         num_scan_states = len(node.input) - num_scan_inputs
         scan_input_axes = [handle_negative_axis(ax, self._get_shape_rank(node, i + num_scan_states)) for i, ax in enumerate(scan_input_axes)]
-        for i, si in enumerate(subgraph.input):
+        # We may have cases where the subgraph has optionial inputs that appear in both subgraph's input and initializer,
+        # but not in the node's input. In such cases, the input model might be invalid, but let's not choke on it.
+        # Instead, let's issue a warning, skip the optional inputs, and keep going forward.
+        subgraph_inputs = subgraph.input
+        if len(subgraph_inputs) != len(node.input):
+            print("warning: skip optionial inputs for scan subgraph {} of scan node {}".format(subgraph.name, node.name))
+            subgraph_initializer_names = [i.name for i in subgraph.initializer]
+            subgraph_inputs = [i for i in subgraph.input if i.name not in subgraph_initializer_names]
```

Comment:
could it be simplified to subgraph_inputs = subgraph_inputs[:len(node.input)]?

Reply:
I think so, otherwise execution will fail because EPs assume 1:1 mapping between Scan node input and subgraph input

---
In reply to: [399605654](https://github.com/microsoft/onnxruntime/pull/3349#discussion_r399605654) [](ancestors = 399605654)

Reply:
That's right. PR updated. Thanks!